src/c-writer.cc

```diff
@@ -2710,20 +2710,32 @@ void CWriter::Write(const ExprList& exprs) {
           Write(StackVar(num_params, decl.GetResultType(0)), " = ");
         }
 
-        const Table* table =
-            module_->GetTable(cast<CallIndirectExpr>(&expr)->table);
-
-        assert(decl.has_func_type);
-        const FuncType* func_type = module_->GetFuncType(decl.type_var);
-
-        Write("CALL_INDIRECT(",
-              ExternalInstanceRef(ModuleFieldType::Table, table->name), ", ");
-        WriteCallIndirectFuncDeclaration(decl, "(*)");
-        Write(", ", FuncTypeExpr(func_type), ", ", StackVar(0));
-        Write(", ", ExternalInstanceRef(ModuleFieldType::Table, table->name),
-              ".data[", StackVar(0), "].module_instance");
+        if (options_.no_sandbox) {
+          Write("((");
+          WriteCallIndirectFuncDeclaration(decl, "(*)");
+          Write(")", StackVar(0), ")(");
+        } else {
+          const Table* table =
+              module_->GetTable(cast<CallIndirectExpr>(&expr)->table);
+
+          assert(decl.has_func_type);
+          const FuncType* func_type = module_->GetFuncType(decl.type_var);
+
+          Write("CALL_INDIRECT(",
+                ExternalInstanceRef(ModuleFieldType::Table, table->name), ", ");
+          WriteCallIndirectFuncDeclaration(decl, "(*)");
+          Write(", ", FuncTypeExpr(func_type), ", ", StackVar(0));
+          Write(", ", ExternalInstanceRef(ModuleFieldType::Table, table->name),
+                ".data[", StackVar(0), "].module_instance");
+          if (num_params > 0) {
+            Write(", ");
+          }
```

Comment:
personal preference: maybe do `needs_comma = true`, and check it inside the loop below? As it is it assumes knowledge of now the code below is written which is generally is undesirable.

Reply:
done